jb/itch5/array_based_order_book.hpp

```diff
@@ -0,0 +1,660 @@
+   * if needed (moving the tail to bottom_levels if any)
+   * - finally, updates qty at the relative(px)
+   */
+  bool add_order(price4_t px, int qty) {
+    if (qty <= 0 or px <= price4_t(0) or
+        px >= max_price_field_value<price4_t>()) {
+      std::ostringstream os;
+      os << "array_based_order_book::add_order value(s) out of range."
+         << " px=" << px << " qty=" << qty;
+      throw jb::feed_error(os.str());
+    }
+    // check if px is worse than the first price of the top_levels_
+    if (better_(px_begin_top_, px)) {
+      // emplace the price at the bottom_levels, and return (false)
+      auto emp_tup = bottom_levels_.emplace(px, 0);
+      emp_tup.first->second += qty;
+      return false;
+    }
+    // check if px is equal or better than the current inside
+    if (not better_(px_inside_, px)) {
+      // ... check if limit redefine is needed
+      if (not better_(px_end_top_, px)) {
+        // get the new limits based on the new inside px
+        auto limits = get_limit_top_prices(px);
+        // move the tail [px_begin_top_, new px_begin_top_) to bottom_levels_
+        move_top_to_bottom(std::get<0>(limits));
+        // ... redefine the limits
+        px_begin_top_ = std::get<0>(limits);
+        px_end_top_ = std::get<1>(limits);
+      }
+      // if the px inside changed, updated
+      if (px_inside_ != px) {
+        px_inside_ = px;
+      }
+      // update top_levels_
+      auto rel_px = price_to_relative(px_inside_);
+      top_levels_.at(rel_px) += qty;
+      return true; // the inside changed
+    }
+    // is a top_levels change different than the inside
+    // udates top_levels_ with new qty
+    auto rel_px = price_to_relative(px);
+    top_levels_.at(rel_px) += qty;
+    return false;
+  }
+
+  /**
+   * Reduce the quantity for a given price.
+   *
+   * @param px the price of the order that was reduced
+   * @param qty the quantity reduced in the order
+   * @returns true if the inside changed
+   *
+   * @throw feed_error qty <= 0
+   * @throw feed_error px should be in bottom_levels_ but is empty
+   * @throw feed_error trying to reduce a non-existing bottom_levels price
+   * @throw feed_error trying to reduce a non-existing top_levels price
+   * @throw feed_error trying to reduce a price better than px_inside_
+   *
+   * Checks and handles if px is a bottom_level_ price
+   * Checks if top_levels_ is empty after reducing
+   * If so, limits have to be redefined, and tail moved into top_levels
+   */
+  bool reduce_order(price4_t px, int qty) {
+    // validates input values, qty has to be a positive number
+    if (qty <= 0) {
+      std::ostringstream os;
+      os << "array_based_order_book::reduce_order value(s) out of range."
+         << " px=" << px << " qty=" << qty;
+      throw jb::feed_error(os.str());
+    }
+
+    // check and handles if it is a bottom_level_ price
+    if (better_(px_begin_top_, px)) {
+      if (bottom_levels_.empty()) {
```

Comment:
If `bottom_levels_` is empty then the find operation around line 258 will also raise an exception.  Why not just remove the first check?  Less code is better, yes, the more descriptive error message might be useful, but this is so rare an event that we do not care, right?